File: example/src/main/java/com/matnsolutions/matlive/lib/widgets/AudioRoomLayout.kt
package com.matnsolutions.matlive.lib.widgets

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Lock
import androidx.compose.material.icons.filled.Person
import androidx.compose.material3.Icon
import androidx.compose.material3.Surface
import androidx.compose.runtime.*
import androidx.compose.runtime.livedata.observeAsState
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.unit.dp
import coil.compose.AsyncImage
import com.matnsolutions.matlive_sdk.audio.define.MatLiveRoomAudioSeat
import com.matnsolutions.matlive_sdk.audio.define.MatLiveUser
import com.matnsolutions.matlive_sdk.audio.mangers.MatLiveRoomManger

@Composable
fun AudioRoomLayout(
    onTakeMic: (Int) -> Unit,
    onMuteMic: (Int) -> Unit,
    onRemoveSpeaker: (Int) -> Unit,
    onLeaveMic: (Int) -> Unit,
    onLockMic: (Int) -> Unit,
    onUnLockMic: (Int) -> Unit,
    onSwitchSeat: (Int) -> Unit,
) {
    val seatService = MatLiveRoomManger.instance.seatService ?: return
    val seats by seatService.seatList.collectAsState()
    val layoutConfig = seatService.layoutConfig ?: return

    Column(modifier = Modifier.fillMaxWidth(),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        layoutConfig.rowConfigs.forEachIndexed { rowIndex, rowConfig ->
            Row(
                modifier = Modifier.padding(bottom = layoutConfig.rowSpacing.dp),
                horizontalArrangement = Arrangement.SpaceAround
            ) {
                for (seatIndex in 0 until rowConfig.count) {
                    val globalIndex = calculateGlobalIndex(rowIndex, seatIndex, layoutConfig)
                    val seat = seats.getOrNull(globalIndex) ?: continue
                    SeatItem(
                        seat = seat,
                        onTakeMic = onTakeMic,
                        onMuteMic = onMuteMic,
                        onRemoveSpeaker = onRemoveSpeaker,
                        onLeaveMic = onLeaveMic,
                        onLockMic = onLockMic,
                        onUnLockMic = onUnLockMic,
                        onSwitchSeat = onSwitchSeat,
                        seatIndex = globalIndex
                    )
                }
            }
        }
    }
}

@Composable
private fun SeatItem(
    seat: MatLiveRoomAudioSeat,
    onTakeMic: (Int) -> Unit,
    onMuteMic: (Int) -> Unit,
    onRemoveSpeaker: (Int) -> Unit,
    onLeaveMic: (Int) -> Unit,
    onLockMic: (Int) -> Unit,
    onUnLockMic: (Int) -> Unit,
    onSwitchSeat: (Int) -> Unit,
    seatIndex: Int
) {
    val isLocked by seat.isLocked.observeAsState(initial = false)
    var showBottomSheet by remember { mutableStateOf(false) }
    Box(modifier = Modifier.padding(horizontal = 6.dp)) {
        Surface(
            modifier = Modifier.size(50.dp),
            shape = CircleShape,
            color = Color.Gray.copy(alpha = 0.3f),
            onClick = {
                showBottomSheet = true
            }
        ) {
            if (isLocked) {
                Icon(Icons.Filled.Lock, contentDescription = "Seat Locked", modifier = Modifier.size(22.dp).align(Alignment.Center))
            } else {
                SeatWidget(seat = seat)
            }
            if (isLocked) {
                Icon(Icons.Filled.Lock, contentDescription = "Seat Locked", modifier = Modifier.size(22.dp).align(Alignment.Center))
            } else {
                SeatWidget(seat = seat)
            }
        }
        if (showBottomSheet) {
            SeatActionBottomSheet(
                user = seat.currentUser.value,
                onTakeMic = { onTakeMic(seatIndex) },
                onMuteMic = { onMuteMic(seatIndex) },
                onRemoveSpeaker = { onRemoveSpeaker(seatIndex) },
                onLeaveMic = { onLeaveMic(seatIndex) },
                onLockMic = { onLockMic(seatIndex) },
                onUnLockMic = { onUnLockMic(seatIndex) },
                onSwitch = { onSwitchSeat(seatIndex) },
                onDismiss = { showBottomSheet = false }
            )
        }
    }
}

@Composable
private fun SeatWidget(seat: MatLiveRoomAudioSeat) {
    val user by seat.currentUser.observeAsState()
    Box(modifier = Modifier.size(50.dp)) {
        // User avatar
        Surface(
            modifier = Modifier.fillMaxSize(),
            shape = CircleShape,
            color = Color.Gray.copy(alpha = 0.3f)
        ) {
            if (user != null) {
                AsyncImage(
                    model = user?.avatarUrlNotifier?.value,
                    contentDescription = "User Avatar",
                    modifier = Modifier
                        .fillMaxSize()
                        .clip(CircleShape),
                    contentScale = ContentScale.Crop,
//                    error = {
//                        Icon(Icons.Filled.Person, contentDescription = "User Avatar", modifier = Modifier.size(22.dp))
//                    }
                )
            } else {
                Icon(Icons.Filled.Check, contentDescription = "Mic", modifier = Modifier.size(22.dp).align(Alignment.Center))
            }
        }
        // Mic icon
        if (user != null) {
            val isMicOn by user!!.isMicOnNotifier.observeAsState(initial = false)
            Box(
                modifier = Modifier
                    .align(Alignment.BottomEnd)
                    .padding(4.dp)
            ) {
                Surface(
                    shape = CircleShape,
                    color = Color.White,
                ) {
                    Icon(
                        imageVector = if (isMicOn) Icons.Filled.Check else Icons.Filled.Close,
                        contentDescription = "Mic Status",
                        modifier = Modifier.size(12.dp),
                        tint = if (isMicOn) Color.Green else Color.Red
                    )
                }
            }
        }
    }
}


private fun calculateGlobalIndex(rowIndex: Int, seatIndex: Int, layoutConfig: com.matnsolutions.matlive_sdk.audio.seats.MatLiveAudioRoomLayoutConfig): Int {
    var totalPreviousSeats = 0
    for (i in 0 until rowIndex) {
        totalPreviousSeats += layoutConfig.rowConfigs[i].count
    }
    return totalPreviousSeats + seatIndex
}

File: example/src/main/java/com/matnsolutions/matlive/lib/widgets/audio_room_layout.dart
import 'package:example/widgets/seat_action_bottom_sheet.dart';
import 'package:flutter/material.dart';
import 'package:matlive_client_sdk_flutter/mat_live.dart';

class AudioRoomLayout extends StatelessWidget {
  final Function(int index)? onTakeMic;
  final Function(int index)? onMuteMic;
  final Function(int index)? onRemoveSpeaker;
  final Function(int index)? onLeaveMic;
  final Function(int index)? onUnLockMic;
  final Function(int index)? onLockMic;
  final Function(int toIndex)? onSwitchSeat;

  const AudioRoomLayout({
    super.key,
    this.onTakeMic,
    this.onMuteMic,
    this.onRemoveSpeaker,
    this.onLeaveMic,
    this.onLockMic,
    this.onUnLockMic,
    required this.onSwitchSeat,
  });

  @override
  Widget build(BuildContext context) {
    if (MatLiveRoomManger.instance.seatService == null) return Container();
    return ValueListenableBuilder<List<MatLiveRoomAudioSeat>>(
      valueListenable: MatLiveRoomManger.instance.seatService!.seatList,
      builder: (context, seats, _) {
        final layoutConfig =
            MatLiveRoomManger.instance.seatService!.layoutConfig!;
        return Column(
          mainAxisSize: MainAxisSize.min,
          children: List.generate(
            layoutConfig.rowConfigs.length,
            (rowIndex) {
              final rowConfig = layoutConfig.rowConfigs[rowIndex];
              return Padding(
                padding: EdgeInsets.only(bottom: layoutConfig.rowSpacing),
                child: Row(
                  mainAxisAlignment: MainAxisAlignment.spaceAround,
                  children: List.generate(
                    rowConfig.count,
                    (seatIndex) {
                      final globalIndex =
                          _calculateGlobalIndex(rowIndex, seatIndex);
                      final seat = seats[globalIndex];
                      return Padding(
                        padding: EdgeInsets.symmetric(
                          horizontal: rowConfig.seatSpacing / 2,
                        ),
                        child: GestureDetector(
                          onTap: () => _showSeatActions(
                            context,
                            globalIndex,
                            user: seat.currentUser.value,
                          ),
                          child: ValueListenableBuilder<bool>(
                            valueListenable: seat.isLocked,
                            builder: (context, isLocked, _) {
                              if (isLocked) {
                                return Container(
                                  width: 50,
                                  height: 50,
                                  decoration: BoxDecoration(
                                    shape: BoxShape.circle,
                                    color: Colors.grey[300],
                                  ),
                                  child: const Icon(Icons.lock, size: 22.0),
                                );
                              }
                              return SeatWidget(
                                seat: seat,
                              );
                            },
                          ),
                        ),
                      );
                    },
                  ),
                ),
              );
            },
          ),
        );
      },
    );
  }

  void _showSeatActions(
    BuildContext context,
    int seatIndex, {
    required MatLiveUser? user,
  }) {
    showModalBottomSheet(
      context: context,
      backgroundColor: Colors.transparent,
      builder: (context) => SeatActionBottomSheet(
        user: user,
        onTakeMic: onTakeMic != null ? () => onTakeMic!(seatIndex) : null,
        onMuteMic: onMuteMic != null ? () => onMuteMic!(seatIndex) : null,
        onRemoveSpeaker:
            onRemoveSpeaker != null ? () => onRemoveSpeaker!(seatIndex) : null,
        onLeaveMic: onLeaveMic != null ? () => onLeaveMic!(seatIndex) : null,
        onLockMic: onLockMic != null ? () => onLockMic!(seatIndex) : null,
        onUnLockMic: onUnLockMic != null ? () => onUnLockMic!(seatIndex) : null,
        onSwitch: onSwitchSeat != null ? () => onSwitchSeat!(seatIndex) : null,
      ),
    );
  }

  int _calculateGlobalIndex(int rowIndex, int seatIndex) {
    final layoutConfig = MatLiveRoomManger.instance.seatService!.layoutConfig!;
    int totalPreviousSeats = 0;
    for (int i = 0; i < rowIndex; i++) {
      totalPreviousSeats += layoutConfig.rowConfigs[i].count;
    }
    return totalPreviousSeats + seatIndex;
  }
}

class SeatWidget extends StatelessWidget {
  final MatLiveRoomAudioSeat seat;

  const SeatWidget({
    super.key,
    required this.seat,
  });

  @override
  Widget build(BuildContext context) {
    return ValueListenableBuilder<MatLiveUser?>(
        valueListenable: seat.currentUser,
        builder: (context, user, _) {
          return SizedBox(
            width: 50,
            height: 50,
            child: Stack(
              children: [
                // User avatar
                Positioned.fill(
                  child: Container(
                    decoration: BoxDecoration(
                      shape: BoxShape.circle,
                      color: Colors.grey[300],
                    ),
                    child: user != null
                        ? ClipOval(
                            child: ValueListenableBuilder<String?>(
                                valueListenable: user.avatarUrlNotifier,
                                builder: (context, avatarUrl, _) {
                                  return Image.network(
                                    '$avatarUrl',
                                    fit: BoxFit.cover,
                                    errorBuilder: (context, error, stackTrace) {
                                      return const Icon(Icons.person,
                                          size: 22.0);
                                    },
                                  );
                                }),
                          )
                        : const Icon(Icons.mic, size: 22.0),
                  ),
                ),
                // Mic icon
                if (user != null)
                  Positioned(
                    bottom: 0,
                    right: 0,
                    child: Container(
                      padding: const EdgeInsets.all(4),
                      decoration: BoxDecoration(
                        color: Colors.white,
                        shape: BoxShape.circle,
                        boxShadow: [
                          BoxShadow(
                            color: Colors.black.withOpacity(0.1),
                            blurRadius: 4,
                            offset: const Offset(0, 2),
                          ),
                        ],
                      ),
                      child: ValueListenableBuilder<bool>(
                          valueListenable: user!.isMicOnNotifier,
                          builder: (context, isMuted, _) {
                            return Icon(
                              isMuted ? Icons.mic : Icons.mic_off,
                              size: 12.0,
                              color: isMuted ? Colors.green : Colors.red,
                            );
                          }),
                    ),
                  ),
              ],
            ),
          );
        });
  }
}
